File: ocaml/day5.ml
open Str

let number_regex = regexp {|\([0-9]+\)|}

let extract_numbers str =
  let rec collect pos acc =
    try
      let _ = search_forward number_regex str pos in
      let num = int_of_string (matched_group 1 str) in
      collect (match_end ()) (num :: acc)
    with Not_found -> List.rev acc
  in
  collect 0 []

let make_seq lines =
  match lines with
  | line :: rest ->
    let numbers = extract_numbers line in
    let remaining = match rest with _ :: t -> t | [] -> [] in (* skip blank line *)
    (numbers, remaining)
  | [] -> ([], [])

let make_map lines =
  let rec skip_empty = function
    | [] -> []
    | "" :: rest -> skip_empty rest
    | lines -> lines
  in
  match skip_empty lines with
  | [] -> None
  | header :: rest ->
    let rec take_while pred acc = function
      | [] -> (List.rev acc, [])
      | x :: xs when pred x -> take_while pred (x :: acc) xs
      | xs -> (List.rev acc, xs)
    in
    let is_data_line s = 
      String.trim s <> "" && not (String.contains s ':')
    in
    let (mapping_lines, remaining) = take_while is_data_line [] rest in
    let ranges = List.map (fun line ->
      match extract_numbers line with
      | [dest; src; len] -> (dest, src, len)
      | nums -> failwith ("Invalid range format: " ^ line ^ " (found " ^ string_of_int (List.length nums) ^ " numbers)")
    ) mapping_lines in
    let map_func i =
      match List.find_opt (fun (_, s, l) -> s <= i && i < s + l) ranges with
      | Some (b, s, _) -> b + i - s
      | None -> i
    in
    Some (map_func, remaining)

let rec get_all_maps lines =
  match make_map lines with
  | None -> []
  | Some (map_func, remaining) -> map_func :: get_all_maps remaining

let compose_functions funcs =
  List.fold_left (fun acc f -> fun x -> f (acc x)) (fun x -> x) funcs

let chunks_of n lst =
  let rec aux acc current = function
    | [] -> if current = [] then List.rev acc else List.rev (List.rev current :: acc)
    | x :: xs ->
      if List.length current = n then
        aux (List.rev current :: acc) [x] xs
      else
        aux acc (x :: current) xs
  in
  aux [] [] lst

let process lines =
  let (seeds, rest) = make_seq lines in
  let all_maps = get_all_maps rest in
  let seed_to_location = compose_functions all_maps in
  
  let part1 = List.fold_left min max_int (List.map seed_to_location seeds) in
  
  let seed_pairs = chunks_of 2 seeds in
  let process_range = function
    | [start; len] ->
        let rec find_min current_min i =
          if i >= start + len then current_min
          else
            let location = seed_to_location i in
            find_min (min current_min location) (i + 1)
        in
        let result = find_min max_int start in
        result
    | lst -> 
      Printf.eprintf "Invalid pair: %s\n" (String.concat "," (List.map string_of_int lst));
      max_int
  in
  let part2 = List.fold_left min max_int (List.map process_range seed_pairs) in
  
  (part1, part2)

let () =
  let lines = 
    let rec read_lines acc =
      try
        let line = read_line () in
        read_lines (line :: acc)
      with End_of_file -> List.rev acc
    in
    read_lines []
  in
  let (part1, part2) = process lines in
  Printf.printf "(%d, %d)\n" part1 part2

File: go/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

var numberRegex = regexp.MustCompile(`\d+`)

func makeSeq(scanner *bufio.Scanner) ([]int64, error) {
	if !scanner.Scan() {
		return nil, fmt.Errorf("no input")
	}
	line := scanner.Text()
	matches := numberRegex.FindAllString(line, -1)
	var numbers []int64
	for _, match := range matches {
		num, err := strconv.ParseInt(match, 10, 64)
		if err != nil {
			return nil, err
		}
		numbers = append(numbers, num)
	}
	// skip blank line
	scanner.Scan()
	return numbers, nil
}

func makeMap(scanner *bufio.Scanner) (func(int64) int64, bool) {
	if !scanner.Scan() {
		return nil, false
	}
	// skip section header
	
	type rangeMapping struct {
		dest, src, length int64
	}
	
	var ranges []rangeMapping
	
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			break
		}
		matches := numberRegex.FindAllString(line, -1)
		if len(matches) != 3 {
			continue
		}
		dest, _ := strconv.ParseInt(matches[0], 10, 64)
		src, _ := strconv.ParseInt(matches[1], 10, 64)
		length, _ := strconv.ParseInt(matches[2], 10, 64)
		ranges = append(ranges, rangeMapping{dest, src, length})
	}
	
	mapFunc := func(i int64) int64 {
		for _, r := range ranges {
			if r.src <= i && i < r.src+r.length {
				return r.dest + i - r.src
			}
		}
		return i
	}
	
	return mapFunc, true
}

func getAllMaps(scanner *bufio.Scanner) []func(int64) int64 {
	var maps []func(int64) int64
	for {
		mapFunc, hasMore := makeMap(scanner)
		if !hasMore {
			break
		}
		maps = append(maps, mapFunc)
	}
	return maps
}

func composeFunctions(funcs []func(int64) int64) func(int64) int64 {
	return func(x int64) int64 {
		result := x
		for _, f := range funcs {
			result = f(result)
		}
		return result
	}
}

func sliding(slice []int64, size, step int) [][]int64 {
	var result [][]int64
	for i := 0; i <= len(slice)-size; i += step {
		window := make([]int64, size)
		copy(window, slice[i:i+size])
		result = append(result, window)
	}
	return result
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

func minSlice(slice []int64) int64 {
	if len(slice) == 0 {
		return 0
	}
	result := slice[0]
	for _, v := range slice[1:] {
		result = min(result, v)
	}
	return result
}

func process(scanner *bufio.Scanner) (int64, int64) {
	seeds, err := makeSeq(scanner)
	if err != nil {
		panic(err)
	}
	
	allMaps := getAllMaps(scanner)
	seedToLocation := composeFunctions(allMaps)
	
	// Part 1
	var part1Results []int64
	for _, seed := range seeds {
		part1Results = append(part1Results, seedToLocation(seed))
	}
	part1 := minSlice(part1Results)
	
	// Part 2
	seedPairs := sliding(seeds, 2, 2)
	var part2Results []int64
	
	for _, pair := range seedPairs {
		start := pair[0]
		length := pair[1]
		
		rangeMin := int64(1<<63 - 1) // max int64
		for i := start; i < start+length; i++ {
			location := seedToLocation(i)
			rangeMin = min(rangeMin, location)
		}
		part2Results = append(part2Results, rangeMin)
	}
	part2 := minSlice(part2Results)
	
	return part1, part2
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	part1, part2 := process(scanner)
	fmt.Printf("(%d, %d)\n", part1, part2)
}
